android-activity/src/intent/mod.rs

```diff
@@ -0,0 +1,183 @@
+    object: JObject<'env>,
+}
+
+/// A messaging object you can use to request an action from another android app component.
+#[must_use]
+pub struct Intent<'env> {
+    inner: Result<Inner<'env>, Error>,
+}
+
+impl<'env> Intent<'env> {
+    pub fn from_object(env: JNIEnv<'env>, object: JObject<'env>) -> Self {
+        Self {
+            inner: Ok(Inner { env, object }),
+        }
+    }
+
+    fn from_fn(f: impl FnOnce() -> Result<Inner<'env>, Error>) -> Self {
+        let inner = f();
+        Self { inner }
+    }
+
+    pub fn new(env: JNIEnv<'env>, action: impl AsRef<str>) -> Self {
+        Self::from_fn(|| {
+            let intent_class = env.find_class("android/content/Intent")?;
+            let action_view =
+                env.get_static_field(intent_class, action.as_ref(), "Ljava/lang/String;")?;
+
+            let intent =
+                env.new_object(intent_class, "(Ljava/lang/String;)V", &[action_view.into()])?;
+
+            Ok(Inner {
+                env,
+                object: intent,
+            })
+        })
+    }
+
+    pub fn new_with_uri(env: JNIEnv<'env>, action: impl AsRef<str>, uri: impl AsRef<str>) -> Self {
+        Self::from_fn(|| {
+            let url_string = env.new_string(uri)?;
+            let uri_class = env.find_class("android/net/Uri")?;
+            let uri = env.call_static_method(
+                uri_class,
+                "parse",
+                "(Ljava/lang/String;)Landroid/net/Uri;",
+                &[JString::from(url_string).into()],
+            )?;
+
+            let intent_class = env.find_class("android/content/Intent")?;
+            let action_view =
+                env.get_static_field(intent_class, action.as_ref(), "Ljava/lang/String;")?;
+
+            let intent = env.new_object(
+                intent_class,
+                "(Ljava/lang/String;Landroid/net/Uri;)V",
+                &[action_view.into(), uri.into()],
+            )?;
+
+            Ok(Inner {
+                env,
+                object: intent,
+            })
+        })
+    }
+
+    /// Add extended data to the intent.
+    /// ```no_run
+    /// use android_intent::{Action, Extra, Intent};
+    ///
+    /// # android_intent::with_current_env(|env| {
+    /// let intent = Intent::new(env, Action::Send);
+    /// intent.push_extra(Extra::Text, "Hello World!")
+    /// # })
+    /// ```
+    pub fn with_extra(self, key: impl AsRef<str>, value: impl AsRef<str>) -> Self {
+        self.and_then(|inner| {
+            let key = inner.env.new_string(key)?;
+            let value = inner.env.new_string(value)?;
+
+            inner.env.call_method(
+                inner.object,
+                "putExtra",
+                "(Ljava/lang/String;Ljava/lang/String;)Landroid/content/Intent;",
+                &[key.into(), value.into()],
+            )?;
+
+            Ok(inner)
+        })
+    }
+
+    /// Builds a new [`Action::Chooser`](Action) Intent that wraps the given target intent.
```

Comment:
```suggestion
    /// Builds a new [`Action::Chooser`] Intent that wraps the given target intent.
```

Let's link directly to the desired enum _variant_ instead of to the type?